src/PostgresOutbox/Subscriptions/Subscription.cs

```diff
@@ -0,0 +1,160 @@
+        }
+        else
+        {
+            slot = new PgOutputReplicationSlot(
+                new ReplicationSlotOptions(
+                    slotSetupOptions.SlotName,
+                    created.LogSequenceNumber
+                )
+            );
+
+            await foreach (var envelope in ReadExistingRowsFromSnapshot(dataSource, created.SnapshotName, _options, ct))
+            await foreach (var subscribe in ProcessEnvelope<IEnvelope>(envelope, registry, errorProcessor).WithCancellation(ct))
+                yield return subscribe;
+        }
+
+        await foreach (var message in
+                       _connection.StartReplication(slot,
+                           new PgOutputReplicationOptions(publicationSetupOptions.PublicationName, 1, slotSetupOptions.Binary), ct))
+        {
+            if (message is InsertMessage insertMessage)
+            {
+                var envelope = await replicationDataMapper.ReadFromReplication(insertMessage, ct);
+                await foreach (var subscribe in ProcessEnvelope<IEnvelope>(envelope, registry, errorProcessor).WithCancellation(ct))
+                    yield return subscribe;
+            }
+            // Always call SetReplicationStatus() or assign LastAppliedLsn and LastFlushedLsn individually
+            // so that Npgsql can inform the server which WAL files can be removed/recycled.
+            _connection.SetReplicationStatus(message.WalEnd);
+            await _connection.SendStatusUpdate(ct);
+        }
+    }
+
+    private static async IAsyncEnumerable<T> ProcessEnvelope<T>(
+        IEnvelope envelope,
+        Dictionary<Type, IConsume> registry,
+        IErrorProcessor errorProcessor
+    ) where T:class
+    {
+        switch (envelope)
+        {
+            case KoEnvelope error:
+                await errorProcessor.Process(error.Error);
+                yield break;
+            case OkEnvelope okEnvelope:
+            {
+                var obj = okEnvelope.Value;
+                var objType = obj.GetType();
+                var (consumer, methodInfo) = Memoize(registry, objType, Consumer);
+                await ((Task)methodInfo.Invoke(consumer, [obj])!).ConfigureAwait(false);
+                yield return (T)envelope;
+                yield break;
+            }
+        }
+    }
+
+    private static readonly Dictionary<Type, (IConsume consumer, MethodInfo methodInfo)> Cache = [];
+    
+
+    private static (IConsume consumer, MethodInfo methodInfo) Memoize
+    (
+        Dictionary<Type, IConsume> registry,
+        Type objType,
+        Func<Dictionary<Type, IConsume>, Type, (IConsume consumer, MethodInfo methodInfo)> func
+    )
+    {
+        if (!Cache.TryGetValue(objType, out var entry))
+            entry = func(registry, objType);
+        Cache[objType] = entry;
+        return entry;
+    }
+    private static (IConsume consumer, MethodInfo methodInfo) Consumer(Dictionary<Type, IConsume> registry, Type objType)
```

Comment:
**SUGGESTION:** Might be in post setup, but I think that we should use here the ImHashMap in the longer term.

Reply:
Never heard of it, thanks. Maybe we can postpone this point

Reply:
Agreed. Added: https://github.com/oskardudycz/PostgresOutboxPatternWithCDC.NET/issues/11